src/validation.cpp

```diff
@@ -5854,3 +5863,40 @@ Chainstate& ChainstateManager::GetChainstateForIndexing()
     return (IsSnapshotActive() && !IsSnapshotValidated()) ?
         *m_ibd_chainstate : *m_active_chainstate;
 }
+
+std::pair<unsigned int, unsigned int> ChainstateManager::GetPruneRange(
+    const Chainstate& chainstate, unsigned int prune_height)
+{
+    if (chainstate.m_chain.Height() <= 0) {
+        return {0, 0};
+    }
+    unsigned int prune_start{0};
+
+    if (this->GetAll().size() > 1) {
+        std::optional<int> snapshot_height{GetSnapshotBaseHeight()};
+
+        // Leave the blocks in the background IBD chain alone if we're pruning
+        // the snapshot chain.
+        if (&chainstate == m_snapshot_chainstate.get() && snapshot_height) {
+            prune_start = *snapshot_height + 1;
+        }
+    }
+
+    unsigned int prune_end{0};
+    unsigned int chain_height{static_cast<unsigned int>(chainstate.m_chain.Height())};
+
+    if (MIN_BLOCKS_TO_KEEP > chain_height) {
+        // Prevent underflow.
+        prune_end = 0;
+    } else {
+        // last block to prune is the lesser of (user-specified height, MIN_BLOCKS_TO_KEEP from the tip)
+        //
+        // While you might be tempted to prune the background chainstate more
+        // aggressively (i.e. fewer MIN_BLOCKS_TO_KEEP), this won't work with index
+        // building - specifically blockfilterindex requires undo data, and if
+        // we don't maintain this trailing window, we hit indexation failures.
+        prune_end = std::min(prune_height, chain_height - MIN_BLOCKS_TO_KEEP);
+    }
```

Comment:
In commit "validation: pruning for multiple chainstates" (b9ac4f7925e2d9dec67be06fdc2a381532581136)

Would suggest compressing this all down to:

```c++
// Set the end of the prune range to the maximum height which is at least
// MIN_BLOCKS_TO_KEEP blocks away from the chain tip, and is not higher than the
// highest block that is allowed to be pruned according to the caller.
prune_end = std::min(std::max<int>(0, chain_height - MIN_BLOCKS_TO_KEEP), last_prune);
```

Assuming `prune_height` is renamed back to `last_prune` as suggested earlier and made into an int.